File: app/src/main/java/com/example/energo_monitoring/checks/adapters/PhotosRecyclerAdapter.java
package com.example.energo_monitoring.checks.adapters;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.activity.result.ActivityResultLauncher;
import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.example.energo_monitoring.checks.activities.CheckMainActivity;
import com.example.energo_monitoring.checks.presenters.utilities.LoadImageManager;
import com.example.energo_monitoring.R;

import java.util.ArrayList;

public class PhotosRecyclerAdapter extends RecyclerView.Adapter<PhotosRecyclerAdapter.PhotoViewHolder> {

    public ArrayList<Bitmap> photos;
    ActivityResultLauncher<Intent> takePhotoLauncher;
    Context context;
    public int listItem;

    public PhotosRecyclerAdapter(Context context, int listItem) {
        this.context = context;
        this.listItem = listItem;

        // create empty element to click it to take first image
        photos = new ArrayList<>();
        photos.add(Bitmap.createBitmap(100, 100, Bitmap.Config.ARGB_8888));
    }

    public void addPhoto(Bitmap bitmapPhoto){
        // insert new photo to end, move photoButton to begin
        photos.remove(photos.size() - 1);
        photos.add(bitmapPhoto);
        photos.add(Bitmap.createBitmap(100, 100, Bitmap.Config.ARGB_8888));
    }

    public void setPhotoLauncher(ActivityResultLauncher<Intent> launcher){
        this.takePhotoLauncher = launcher;
    }

    public static class PhotoViewHolder extends RecyclerView.ViewHolder{

        public PhotoViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        public void loadImageFromBitmap(Bitmap bitmap){
            ImageView photoImageView = itemView.findViewById(R.id.photoImageView);
            photoImageView.setImageBitmap(bitmap);
        }
    }

    @NonNull
    @Override
    public PhotoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(listItem, parent, false);

        PhotoViewHolder vh = new PhotoViewHolder(view);

        CardView cardView = view.findViewById(R.id.photoCard);
        cardView.setOnClickListener(v -> {
            if(vh.getLayoutPosition() == photos.size() - 1) {
                ((CheckMainActivity) context).setLastCreatedPath(LoadImageManager.getPhotoUri(context));
                LoadImageManager.takePhoto(
                        context,
                        takePhotoLauncher,
                        ((CheckMainActivity) context).getLastCreatedPath());
            }
        });

        return vh;
    }

    @Override
    public void onBindViewHolder(@NonNull PhotoViewHolder holder, int position) {
        //if(position != photos.size()){
            holder.loadImageFromBitmap(photos.get(position));
        //}
    }

    @Override
    public int getItemCount() {
        return photos.size();
    }

}

File: app/src/main/java/com/example/energo_monitoring/checks/ui/adapters/PhotosRecyclerAdapter.kt
package com.example.energo_monitoring.checks.ui.adapters

import android.content.Context
import com.example.energo_monitoring.checks.ui.presenters.utilities.LoadImageManager.Companion.getPhotoUri
import com.example.energo_monitoring.checks.ui.presenters.utilities.LoadImageManager.Companion.takePhoto
import androidx.recyclerview.widget.RecyclerView
import android.graphics.Bitmap
import androidx.activity.result.ActivityResultLauncher
import android.content.Intent
import com.example.energo_monitoring.R
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import androidx.cardview.widget.CardView
import com.example.energo_monitoring.checks.ui.activities.CheckMainActivity
import java.util.ArrayList

class PhotosRecyclerAdapter(var context: Context, var listItem: Int) :
    RecyclerView.Adapter<PhotosRecyclerAdapter.PhotoViewHolder>() {
    var photos: ArrayList<Bitmap> = ArrayList()
    var takePhotoLauncher: ActivityResultLauncher<Intent>? = null
    fun addPhoto(bitmapPhoto: Bitmap) {
        // insert new photo to end, move photoButton to begin
        photos.removeAt(photos.size - 1)
        photos.add(bitmapPhoto)
        photos.add(Bitmap.createBitmap(100, 100, Bitmap.Config.ARGB_8888))
    }

    fun setPhotoLauncher(launcher: ActivityResultLauncher<Intent>?) {
        takePhotoLauncher = launcher
    }

    class PhotoViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun loadImageFromBitmap(bitmap: Bitmap?) {
            val photoImageView = itemView.findViewById<ImageView>(R.id.photoImageView)
            photoImageView.setImageBitmap(bitmap)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PhotoViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(listItem, parent, false)
        val vh = PhotoViewHolder(view)
        val cardView: CardView = view.findViewById(R.id.photoCard)
        cardView.setOnClickListener { v: View? ->
            if (vh.layoutPosition == photos.size - 1) {
                (context as CheckMainActivity).lastCreatedPath = getPhotoUri(context) {
                    (context as CheckMainActivity).lastCreatedPathReal = it
                }
                takePhoto(takePhotoLauncher!!, (context as CheckMainActivity).lastCreatedPath)
            }
        }
        return vh
    }

    override fun onBindViewHolder(holder: PhotoViewHolder, position: Int) {
        holder.loadImageFromBitmap(photos[position])
    }

    override fun getItemCount(): Int {
        return photos.size
    }

    init {

        // create empty element to click it to take first image
        photos.add(Bitmap.createBitmap(100, 100, Bitmap.Config.ARGB_8888))
    }
}